hopsworks-common/src/main/java/io/hops/hopsworks/common/user/security/apiKey/ApiKeyController.java

```diff
@@ -0,0 +1,356 @@
+        "Failed to generate unique key prefix after " + RETRY_KEY_CREATION + " retries.");
+    }
+    return secret;
+  }
+  
+  private List<ApiKeyScope> getKeyScopes(Set<ApiScope> scopes, ApiKey apiKey) {
+    List<ApiKeyScope> keyScopes = new ArrayList<>();
+    for (ApiScope scope : scopes) {
+      keyScopes.add(new ApiKeyScope(scope, apiKey));
+    }
+    return keyScopes;
+  }
+  
+  /**
+   *
+   * @param apiKey
+   * @return
+   */
+  public Set<ApiScope> getScopes(ApiKey apiKey) {
+    Set<ApiScope> scopes = new HashSet<>();
+    for (ApiKeyScope scope : apiKey.getApiKeyScopeCollection()) {
+      scopes.add(scope.getScope());
+    }
+    return scopes;
+  }
+  
+  /**
+   *
+   * @param key
+   * @return
+   * @throws ApiKeyException
+   */
+  public ApiKey getApiKey(String key) throws ApiKeyException {
+    String[] parts = key.split(Secret.KEY_ID_SEPARATOR_REGEX);
+    if (parts.length < 2) {
+      throw new ApiKeyException(RESTCodes.ApiKeyErrorCode.KEY_NOT_FOUND, Level.FINE);
+    }
+    ApiKey apiKey = apiKeyFacade.findByPrefix(parts[0]);
+    if (apiKey == null) {
+      throw new ApiKeyException(RESTCodes.ApiKeyErrorCode.KEY_NOT_FOUND, Level.FINE);
+    }
+    //___MinLength can be set to 0 b/c no validation is needed if the key was in db
+    Secret secret = new Secret(parts[0], parts[1], apiKey.getSalt());
+    if (!secret.getSha256HexDigest().equals(apiKey.getSecret())) {
+      throw new ApiKeyException(RESTCodes.ApiKeyErrorCode.KEY_NOT_FOUND, Level.FINE);
+    }
+    return apiKey;
+  }
+  
+  /**
+   *
+   * @param user
+   * @return
+   */
+  public List<ApiKey> getKeys(Users user) {
+    return apiKeyFacade.findByUser(user);
+  }
+  
+  /**
+   *
+   * @param user
+   * @param keyName
+   */
+  public void deleteKey(Users user, String keyName, HttpServletRequest req) {
+    ApiKey apiKey = apiKeyFacade.findByUserAndName(user, keyName);
+    if (apiKey == null) {
+      return;
+    }
+    apiKeyFacade.remove(apiKey);
+    sendDeletedEmail(user, keyName, req);
+  }
+  
+  /**
+   *
+   * @param user
+   */
+  public void deleteAll(Users user, HttpServletRequest req) {
+    List<ApiKey> keys = apiKeyFacade.findByUser(user);
+    for (ApiKey key : keys) {
+      apiKeyFacade.remove(key);
+    }
+    sendDeletedAllEmail(user, req);
+  }
+  
+  /**
+   *
+   * @param user
+   * @param keyName
+   * @param scopes
+   * @return
+   * @throws ApiKeyException
+   */
+  public ApiKey addScope(Users user, String keyName, Set<ApiScope> scopes) throws ApiKeyException {
+    ApiKey apiKey = validate(user, keyName, scopes);
+    List<ApiKeyScope> newScopes;
+    Set<ApiScope> oldScopes = toApiScope(apiKey.getApiKeyScopeCollection());
+    scopes.removeAll(oldScopes);
+    if (!scopes.isEmpty()) {
+      newScopes = getKeyScopes(scopes, apiKey);
+      apiKey.getApiKeyScopeCollection().addAll(newScopes);
+      apiKey.setModified(new Timestamp(new Date().getTime()));
```

Comment:
Does it need to be wrapped around Timestamp?

Reply:
I was getting expected timestamp but got date exception but I can test it again